ariadne/cardano/src/Ariadne/Wallet/Backend/Util.hs

```diff
@@ -0,0 +1,103 @@
+module Ariadne.Wallet.Backend.Util
+       ( allAccountIds
+       , ourAccountIdFromAccounts
+       , accountsToUse
+       ) where
+
+import Control.Lens (ix)
+import qualified Data.Text.Buildable
+import Formatting (bprint, build, int, (%))
+
+import Pos.Util (maybeThrow)
+
+import Ariadne.Wallet.Backend.KeyStorage (resolveWalletRefThenRead)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+  (HdAccount, HdAccountId, HdRootId, hdAccountId, hdAccountIdParent)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Read (readAccountsByRootId)
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet (IxSet, (@+))
+import Ariadne.Wallet.Cardano.WalletLayer (PassiveWalletLayer(..))
+import Ariadne.Wallet.Face
+  (LocalAccountReference(..), WalletReference, WalletSelection(WSAccount))
+
+
+data OurAccountException = OurAccountNoAccounts !HdRootId
+                         | OurAccountIndexOutOfRange !HdRootId !Word
+                       deriving (Show)
+
+instance Buildable OurAccountException where
+    build = \case
+        OurAccountNoAccounts walletIdx ->
+            bprint ("Wallet "%build%" does not have any accounts") walletIdx
+        OurAccountIndexOutOfRange rootId idx ->
+            bprint ("Account #"%int%" doesn't exist in "%build) idx rootId
+
+instance Exception OurAccountException where
+    displayException = toString . prettyL
+
+-- | Used for fee estimate
+allAccountIds :: PassiveWalletLayer IO
+             -> IORef (Maybe WalletSelection)
+             -> WalletReference
+             -> [LocalAccountReference]
+             -> IO [HdAccountId]
+allAccountIds pwl walletSelRef walletRef accRefs =
+         fmap (^. hdAccountId) . toList . snd
+     <$> accountsToUse pwl walletSelRef walletRef accRefs
+
+-- | Used for sending tx with existing accountsToUse
+ourAccountIdFromAccounts :: MonadThrow m
+                         => HdRootId
+                         -> IxSet HdAccount
+                         -> m HdAccountId
+ourAccountIdFromAccounts walletRootId accountsToUse' = do
+    ourAccount <-
+        maybeThrow
+            (OurAccountNoAccounts walletRootId)
+            (toList accountsToUse'  ^? ix 0)
+    return $ ourAccount ^. hdAccountId
+
+accountsToUse :: PassiveWalletLayer IO
+              -> IORef (Maybe WalletSelection)
+              -> WalletReference
+              -> [LocalAccountReference]
+              -> IO (HdRootId, IxSet HdAccount)
+accountsToUse pwl walletSelRef walletRef accRefs = do
+    -- We pick one of the accounts to generate change address in it.
```

Comment:
This comment is in the wrong place. We pick one of the accounts in `ourAccountIdFromAccounts`.